### ISSUE TITLE
[RLC] how to make apiVersion work if there's more than one apiVersion parameters in the input files

### ISSUE THREAD

**Issue by qiaozha** [Azure/autorest.typescript#1565]
We are trying to release rlc for arm-compute for portal team, but I found in compute's latest tag, they have mixed api versions in their default tag of input files. 
https://github.com/Azure/azure-rest-api-specs/tree/main/specification/compute/resource-manager#tag-package-2022-07-02

previously, we generate RLC like 
`  options.apiVersion = options.apiVersion ?? "2021-03-01";`
in the createClient function,  this will work only for one api version in the input files.

In high level client, we have something like OperationSpec to specify the api version parameter, which can link to the api version parameter definition of that specific operation https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/compute/arm-compute/src/operations/disks.ts#L700.

In RLC's case, I wonder If we could generate an apiVersion in the queryParameters of each operation with a default value and make that apiVersion parameter optional to make it work ? 

@joheredi @xirzec @bterlson Could you share your thoughts about it ? Thanks

**Comment by xirzec** [comment]
If I understand this correctly, the generator uses a compound version token like `package-2022-07-02` which is translated to specific api-versions for particular operations, e.g. `2022-03-01` for compute paths?

So rather than having a global api-version, we have to generate the default api-version per path. Probably the simplest answer is as you suggest. I don't think we'd gain much from being clever and trying to support the "profile" or whatever we call it when we have mixed versions in a single tag.

Curious what @joheredi thinks about this one though.

**Comment by qiaozha** [comment]
The experience is that if all the input files are using the same api version.  The generated rlc library can be used like 
```
 const result = await client.path(
    "/subscriptions/{subscriptionId}/providers/Microsoft.ContainerService/managedClusters",
    subscriptionId
  ).get();
```
And we don't need to pass api version at all, because we have set the default client level api version here.
`options.apiVersion = options.apiVersion ?? "2021-03-01";`
In this case, there're two ways for the customers to modify the api version if they like. 
The first one is that we pass a api version when create the client.
```
  const client = ContainerServiceManagementClient(credential, {apiVersion: "2023-01-01"});
  const result = await client.path(
    "/subscriptions/{subscriptionId}/providers/Microsoft.ContainerService/managedClusters",
    subscriptionId
  ).get();
```
The second one is that we give a different value to the operation.
```
  const result = await client.path(
    "/subscriptions/{subscriptionId}/providers/Microsoft.ContainerService/managedClusters",
    subscriptionId
  ).get({
    queryParameters: {
      "api-version": "2022-11-01"
    }
  });
```
we will be able to allow customers to pass a api version at operation level. 

Please note that, currently, if we set the api version in both ways, the operation level will take effect. @joheredi 

But if there are mixed api versions in the input files, we won't generate the above line. And the customer has to pass api version in each call  like 
```
const result = await client.path(
  "/subscriptions/{subscriptionId}/providers/Microsoft.ContainerService/managedClusters",
  subscriptionId
).get({
  queryParameters: {
    "api-version": "2021-03-01"
  }
});
```
here, we can't allow customer to change the value of api-version. because we have defined the type of api-version in the query parameter definition like 
```
export interface ManagedClustersListQueryParamProperties {
  /** Api Version */
  "api-version": "2021-03-01";
}
```
And because we don't have a client level api version, we also don't allow customer to modify the api version as the first way in previous case. 

I feel like the experience for mixed api version's case doesn't look very user friendly. wonder if there's any way to make it better.

**Comment by qiaozha** [comment]
> Please note that, currently, if we set the api version in both ways, the operation level will take effect. @joheredi

Just found out if we set both client level parameter and operation query parameter for api version like 
 ```
  const client = ContainerServiceManagementClient(credential, {apiVersion: "2023-01-01"});
  const result = await client.path(
    "/subscriptions/{subscriptionId}/providers/Microsoft.ContainerService/managedClusters",
    subscriptionId
  ).get({
    queryParameters: {
      "api-version": "2022-11-01"
    }
  });
```
either one will work, the reason it worked before is because we have set some [customizedApiVersion policy](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/containerservice/arm-containerservice-rest/src/containerServiceClient.ts#L36-L40) in @azure-rest/arm-containerservice package before in order to by pass their service backend returns a different api version for long running operation polling request.  

if I remove the above customizeApiVersion logic, I found the final url is 
`https://management.azure.com/subscriptions/subid/providers/Microsoft.ContainerService/managedClusters?api-version=2022-11-01&api-version=2023-01-01`
And the error we get from ARM is 
The resource type 'managedClusters' could not be found in the namespace 'Microsoft.ContainerService' for api version '2022-11-01,2023-01-01'. The supported api-versions are

**Comment by MaryGao** [comment]
The [Azure SDK Guidelines](https://azure.github.io/azure-sdk/general_design.html#general-service-apiversion-7) say:
> ✅ DO add or replace the api-version query parameter on any URI returned by the service e.g., Operation-Location, next page links, etc., with the service version passed configured on the client.

I'd like to summarize our current situation and we have several apiVersion

| Type | Level | Value Source Type| RLC| Comments|
| --------- | -------------- | --------------- |--------------- |--------------- |
 | Client | Client | Default | Supported | When `count(apiVersion) = 1` in input file, we set it as default|
| Client | Client | Customized | Supported | User could customize global apiVersion |
| Client | Operation | Default | Not Implement | When `count(apiVersion) > 1`, we didn't set default value for operations |
| Client | Operation | Customized | Supported | When `count(apiVersion) = 1`, users could choose to ignore this or not;   <br>When `count(apiVersion) > 1`, users have to set the default apiVesion of that operation |
| Service | Operation | Default | Supported | Happened in LRO or pagination links and returned from server side |

According to above table we need to decide

- Do you have a light way to support the default apiVersion at operation level?
- What is their precedence cross different types?

So the current implementations may not match all our assumptions and per Qiaoqiao's descriptions I could notice below existing issues:

- The current implementation may not be user-friendly
- The service may not follow the guideline to support latest apiVersion in polling or paging so we need to do [customizedApiVersion policy](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/containerservice/arm-containerservice-rest/src/containerServiceClient.ts#L36-L40)
- The current core [api policy](https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-client-rest/src/apiVersionPolicy.ts) may have bugs when we have apiVersion in client level and operation level

**Comment by joheredi** [comment]
**tl;dr; -** From a Rest Client perspective the current is the right experience, we are true to the intent and what the REST API describes, while still helping users through IntelliSense.

**Full story**
I chatted with @bterslon about this offline. We think that the current experience where:

- No default client level api-version is set
- Users are required to set the api-version on each operation
- Intellisense guides the user to set the api-version correctly.

Is the right UX from a rest client perspective, this will be pretty much the same thing someone consuming the REST API directly will do, but with the help of Typescript and Intellisense, we are still guiding our users to success.

If users want to use an "unsupported" combination of api-version and operation, the right way to do so would be through the `client.pathUnchecked` method which gives full control of the request.

For the case where both client and operation api-versions are being set, it looks like a bug in @azure-rest/core-client we need to make sure that `operation` takes precedence. @qiaozha would you mind filing an issue for this?
